### ISSUE TITLE
Awesome 4.3-1 doesn't start properly when using xrandr in .xinitrc

### ISSUE THREAD

**Issue by furgerf** [awesomeWM/awesome#2697]
Hi there,

I'm running AwesomeWM on ArchLinux on a laptop with an external display. In .xinitrc, I make sure that this external display is enabled to extend the primary display. (Without invoking xrandr, the external display duplicates the primary display.)

The following xrandr commands are executed:

```sh
for out in $ENABLED; do
  if (xrandr | grep "$out disconnected"); then
    xrandr --output "$out" --off
  fi
done
```

and

```sh
  xrandr --output "$IN" --auto --primary --set "scaling mode" "Full aspect" --pos 0x1085 \
    --output "$EXT" --auto
```

After an update from 4.2-2 to 4.3-1, I only see a gray screen after starting Awesome, all I seem to be able to do is move the cursor. There's no menubar, right-click doesn't open the menu, and the keyboard shortcuts don't work.

I never had this issue prior to 4.3 and I'm afraid I don't have useful logs to add. Please tell me in can provide additional info. For now, I've downgraded to 4.2 where everything works for me as expected.

Output of `awesome --version`:

awesome v4.3 (Too long)
 • Compiled against Lua 5.3.5 (running with Lua 5.3)
 • D-Bus support: ✔
 • execinfo support: ✔
 • xcb-randr version: 1.6
 • LGI version: 0.9.2

**How to reproduce the issue:**

Invoke xrandr before starting awesome.

Alternatively: invoke xrandr before re-starting awesome. This also happens when awesome is merely restarted, i.e.:
- Start awesome without changing the displays -> everything's fine, except that the display is duplicated
- Use xrandr to change the display config -> the displays are fine, but the external display has no menubar
- Re-start awesome to create menubars on all displays (this soft restart) -> same issue as described above, except that the screens aren't gray but show the previous content.

**Comment by psychon** [comment]
> and I'm afraid I don't have useful logs to add. Please tell me in can provide additional info.

Uhm, okay... So, some quick random thoughts:

* Does starting e.g. xterm instead of AwesomeWM after xrandr work fine? Or alternatively: Another WM instead of AwesomeWM? (= Is there a problem with the X11 server instead of with AwesomeWM?)
* Does this also happen with the default config instead of your own custom one?
* Does a `print("I live!")` at the *beginning* of `rc.lua` show up in the logs? (Whichever logs that is depends on how you start AwesomeWM)
* Does a `print("I live!")` at the *end* of `rc.lua` show up in the logs? (Does something get stuck in `rc.lua`?)
* Does a `gears.timer.start_new(10, function() print("still alive") end)` show up in the log?
* Perhaps all of the above with default config vs. your own config? Feel free to test most of these things at the same time, i.e. all the `print`s can be added to the config at once.

So much for my "does something get stuck somewhere?".

> * the displays are fine, but the external display has no menubar

Uhm. Does this mean that AwesomeWM does not see/use some display even though it shows up in the output of `xrandr`? Somehow I now feel unsure which symptoms you see/don't see when you do what.

**Comment by furgerf** [comment]
Thanks for your response, now I think I have some more useful details :)

> Does starting e.g. xterm instead of AwesomeWM after xrandr work fine? Or alternatively: Another WM instead of AwesomeWM? (= Is there a problem with the X11 server instead of with AwesomeWM?)

Yes, everything else works as far as I can tell. I only have problems with AwesomeWM 4.3 - if you consider AwesomeWM 4.2 to be another WM, then yes, other WMs work (though I didn't try others).

> * Does this also happen with the default config instead of your own custom one?
> * Does a `print("I live!")` at the _beginning_ of `rc.lua` show up in the logs? (Whichever logs that is depends on how you start AwesomeWM)
> * Does a `print("I live!")` at the _end_ of `rc.lua` show up in the logs? (Does something get stuck in `rc.lua`?)
> * Does a `gears.timer.start_new(10, function() print("still alive") end)` show up in the log?
> * Perhaps all of the above with default config vs. your own config? Feel free to test most of these things at the same time, i.e. all the `print`s can be added to the config at once.

I should've said that before, my config is [here](https://github.com/furgerf/dotfiles/blob/master/awesome/rc.lua). With the print statements, I narrowed it down to the creation of my widgets, which I have [here](https://github.com/furgerf/dotfiles/blob/master/awesome/widgets.lua). If I don't create the widgets, AwesomeWM 4.3 starts fine.

Since the widgets use vicious (I have 2.3.1 installed), is it possible that there's now an incompatibility which causes the freeze? How can I investigate this further?

> > * the displays are fine, but the external display has no menubar
> 
> Uhm. Does this mean that AwesomeWM does not see/use some display even though it shows up in the output of `xrandr`? Somehow I now feel unsure which symptoms you see/don't see when you do what.

I'm sorry, that was wrong: I do have a menubar, but I don't have any tags on that screen (I use tyrannical). I didn't take the time to make sure that the tags are the way I want them when changing the number of monitors, since I do that very rarely, so I "fix" them by with that soft AwesomeWM restart.

My point was that the problem occurs both when starting a new instance of AwesomeWM as well as when re-starting an existing instance.

**Comment by psychon** [comment]
> With the print statements, I narrowed it down to the creation of my widgets, which I have [here](https://github.com/furgerf/dotfiles/blob/master/awesome/widgets.lua). If I don't create the widgets, AwesomeWM 4.3 starts fine.

So, does that mean that `print` at the beginning works fine, but `print` at the end is not shown? And you narrowed it down by adding more `print`s? If so, can't you find the specific place where things hang by adding more `print`s?

> Since the widgets use vicious (I have 2.3.1 installed), is it possible that there's now an incompatibility which causes the freeze?

Well, possible of course. However, I cannot think of a specific problem right now.

One more debugging idea: You could switch to another VT when things are stuck and look at output of `top` and `pstree $(pidof awesome)`. Does AwesomeWM use 100% CPU? Is it perhaps stuck on some child process?

In case of 100% CPU: Perhaps `strace -p $(pidof awesome)` has something useful, but I doubt it.

**Comment by Elv13** [comment]
As usual with these errors, please read https://awesomewm.org/apidoc/libraries/awful.spawn.html and the "There is a noticeable delay with inputs" entry in the [FAQ](https://awesomewm.org/apidoc/documentation/90-FAQ.md.html). Using `io.open()` and `os.execute()` is highly discouraged and well known to cause this. It may not even be due to the Awesome upgrade, but any package upgrade. Favor using the `easy_async` async method or `awful.widget.watch` when possible.

**Comment by psychon** [comment]
> it may not even be due to the Awesome upgrade, but any package upgrade

vs

> For now, I've downgraded to 4.2 where everything works for me as expected.

**Comment by furgerf** [comment]
> So, does that mean that `print` at the beginning works fine, but `print` at the end is not shown? And you narrowed it down by adding more `print`s? If so, can't you find the specific place where things hang by adding more `print`s?

Yes, that's what I meant. Furthermore, if I comment out the creation of my widgets (which use vicious) then everything works fine. As a test, I only create the volume widget. If I don't register to the vicious events (commenting out [this](https://github.com/furgerf/dotfiles/blob/master/awesome/widgets.lua#L60-L75)), it works but if this is there, AwesomeWM hangs.

> One more debugging idea: You could switch to another VT when things are stuck and look at output of `top` and `pstree $(pidof awesome)`. Does AwesomeWM use 100% CPU? Is it perhaps stuck on some child process?

Nope, there's no high CPU usage. If anything, I'd say `awesome` is using less than usual (but don't quote me on that...).

Something else: I verified that the problem only occurs when _enabling_ the external display. If I merely turn it off, then everything works as expected.

> As usual with these errors, please read https://awesomewm.org/apidoc/libraries/awful.spawn.html and the "There is a noticeable delay with inputs" entry in the [FAQ](https://awesomewm.org/apidoc/documentation/90-FAQ.md.html). Using `io.open()` and `os.execute()` is highly discouraged and well known to cause this. It may not even be due to the Awesome upgrade, but any package upgrade. Favor using the `easy_async` async method or `awful.widget.watch` when possible.

Yes, I'm sure there are many things I could improve in my config. However, at the moment I have neither a good reason nor the time to do this - I currently don't see this being related to my problem but I'm open to be shown to be wrong :)

**Comment by psychon** [comment]
You forgot to check output of `pstree`, I think. The idea here is basically the same as what follows: Is awesome stuck waiting on some child process? The following is another way to check if this is blocked due to `io.popen`.

> If I don't register to the vicious events (commenting out [this](https://github.com/furgerf/dotfiles/blob/master/awesome/widgets.lua#L60-L75)), it works but if this is there,

Okay, that narrows it down. I guess this means that this code here is the problem:
https://github.com/vicious-widgets/vicious/blob/master/widgets/volume_linux.lua#L22-L59

Could you try adding the following to the beginning of your `rc.lua` and see if this really is the problem? (This is basically what the volume widget does, the rest is some "messing around with strings")
```lua
print("before")
local f = io.popen("amixer -M get Master")
print(f:read("*all"))
f:close()
print("after (if you see this, this means the hang is somewhere else)")
```
However, I have no idea why this would only break when the second display is connected. My only guess would be "something pulseaudio". More specifically: It could be that pulseaudio is automatically started via amixer and somehow the X11 server is involved in this, but blocked by awesome still starting up. Or alternatively, DBus gets autostarted and also locks up by waiting for the WM somehow? I don't know, but there were similar problems before, which is the reason why we tell people not to use `io.popen`.

**Comment by furgerf** [comment]
Yes I forgot to post that, sorry:

```
awesome-+-amixer
        |-dropbox---83*[{dropbox}]
        |-find
        |-nm-applet---4*[{nm-applet}]
        |-thunar---2*[{thunar}]
        `-2*[{awesome}]
```

dropbox, nm-applet, and thunar are started from .xinitrc, I'm not sure where the find comes from...

I've added your snippet, that doesn't cause any problems - as long as I run it before `hints.init` [here](https://github.com/furgerf/dotfiles/blob/master/awesome/rc.lua#L54). If I run it afterwards, it freezes. What are your thoughts on that?

I've also checked that only 2 of the widgets cause the freeze: volume and caps lock. The latter is something that I built to show when caps lock is pressed. That could be what @Elv13 was referring to: https://github.com/furgerf/dotfiles/blob/master/awesome/widgets.lua#L482.

> However, I have no idea why this would only break when the second display is connected. My only guess would be "something pulseaudio". More specifically: It could be that pulseaudio is automatically started via amixer and somehow the X11 server is involved in this, but blocked by awesome still starting up. Or alternatively, DBus gets autostarted and also locks up by waiting for the WM somehow? I don't know, but there were similar problems before, which is the reason why we tell people not to use `io.popen`.

I'm not sure what to do with this, I'm really not familiar with that stuff, sorry. I don't think my specific usage of vicious' volume widget is wrong (as I'm not spawning anything or opening pipes myself) so maybe this would be something that needs to be fixed there? I see they've just released a bugfix version though I don't see anything related to my problem in the [changelog](https://github.com/vicious-widgets/vicious/releases).

**Comment by psychon** [comment]
Okay, so this is definitely something related to `io.popen`. I do not know why this hangs and I especially do not know why this only hangs when the external monitor is enabled. All my theories so far would mean "it must always hang" (e.g. commit 4c322a613f508 could be one reason).

I reported this to vicious: https://github.com/vicious-widgets/vicious/issues/74

**Comment by psychon** [comment]
> I'm not sure where the find comes from...

Uhm. I would guess it comes from here: https://github.com/furgerf/dotfiles/blob/d537a72a44018bf06a1ba486032ecae09cead42a/awesome/theme/theme.lua#L32
This code does not close the file handle, thus the C function `pclose` is not called and the zombie process is not reaped. Once AwesomeWM's main loop runs, it will reap the zombie child.

So, to me this means that `io.popen` basically works and the problem here must be `amixer` hanging for some reason that is beyond me.

**Comment by reinerh** [comment]
Could it be that a second monitor registers another (HDMI) sound device, which amixer then tries to use (but fails)?

**Comment by psychon** [comment]
@reinerh https://github.com/awesomeWM/awesome/issues/2697#issuecomment-467404489

> I've added your snippet, that doesn't cause any problems - as long as I run it before `hints.init` [here](https://github.com/furgerf/dotfiles/blob/master/awesome/rc.lua#L54). If I run it afterwards, it freezes. What are your thoughts on that?

The above is highly confusing. Perhaps this is just a timing issue and "doing something slow" is necessary for the HDMI device to show up and cause problems? But I find that unlikely since the screen does seem to work.

(off topic mode activated)
@reinerh This is an automatic reply that I got when trying to reply to your recent mail:

> This is the mail system at host [my mail host]
> 
> I'm sorry to have to inform you that your message could not
>  be delivered to one or more recipients. It's attached below.
> [...]
> 
> connect to mail.reiner-h.de[79.140.41.118]:25: Connection timed out

**Comment by reinerh** [comment]
(offtopic)
@psychon Hm, interesting. Not sure what is wrong there, as I'm receiving a lot of other mails.
Is this a pure mail host, or are you able to check if other ports (ssh, http(s)) are not reachable as well?

**Comment by Elv13** [comment]
> That could be what @Elv13 was referring to:

> Yes, I'm sure there are many things I could improve in my config. However, at the moment I have neither a good reason nor the time to do this - I currently don't see this being related to my problem but I'm open to be shown to be wrong :)

My point is that it is unsupported and unsupportable. Any upgrade to either AwesomeWM or any other commands that:

 * Call an hardware device
 * Perform queued I/O or async system calls.
 * Use dbus [1]
 * Sleep
 * Contact the window manager directly or indirectly (all GUI apps)

Can cause a regression if you use `io.popen` at any moment. Those may or may not make any sense atomically. If your hardware is under I/O load, it will induce very bad input latency that is 100% avoidable. It will never work properly and will always cause a stream of bug reports that can be summarized to "please stop doing that and if the problem persist, we will take another look".

[1] which may signal some other apps to wake up, call the wm, freeze and propagate frozen states into some services attached to them using the synchronous dbus API, freeze *those* then freeze the original command because it tries to access those services later on. And yes, I it happens, pulseaudio<->MPRIS2 are prime examples.

**Comment by furgerf** [comment]
@psychon @reinerh: The monitor is connected via VGA. Full `xrandr` output (this is the situation where the freeze occurs):
```
Screen 0: minimum 8 x 8, current 1920 x 1985, maximum 32767 x 32767
LVDS1 connected primary 1600x900+0+1085 (normal left inverted right x axis y axis) 310mm x 170mm
   1600x900      60.01*+  59.82    40.00  
   1400x900      59.96    59.88  
   1368x768      60.00    59.88    59.85  
   1280x800      59.81    59.91  
   1280x720      59.86    60.00    59.74  
   1024x768      60.00  
   1024x576      60.00    59.90    59.82  
   960x540       60.00    59.63    59.82  
   800x600       60.32    56.25  
   864x486       60.00    59.92    59.57  
   800x450       60.00  
   640x480       59.94  
   720x405       59.51    60.00    58.99  
   640x360       59.84    59.32    60.00  
DP1 disconnected (normal left inverted right x axis y axis)
DP2 disconnected (normal left inverted right x axis y axis)
DP3 disconnected (normal left inverted right x axis y axis)
HDMI1 disconnected (normal left inverted right x axis y axis)
HDMI2 disconnected (normal left inverted right x axis y axis)
HDMI3 disconnected (normal left inverted right x axis y axis)
VGA1 connected 1920x1080+0+0 (normal left inverted right x axis y axis) 520mm x 290mm
   1920x1080     60.00*+
   1680x1050     59.95  
   1280x1024     75.02    60.02  
   1440x900      59.89  
   1280x960      60.00  
   1280x720      60.00  
   1024x768      75.03    70.07    60.00  
   832x624       74.55  
   800x600       72.19    75.00    60.32    56.25  
   640x480       75.00    72.81    66.67    59.94  
   720x400       70.08  
VIRTUAL1 disconnected (normal left inverted right x axis y axis)
```

And `cat /proc/asound/cards`:
```
 0 [PCH            ]: HDA-Intel - HDA Intel PCH
                      HDA Intel PCH at 0xf2530000 irq 31
```

> Okay, so this is definitely something related to `io.popen`. I do not know why this hangs and I especially do not know why this only hangs when the external monitor is enabled. All my theories so far would mean "it must always hang" (e.g. commit [4c322a6](https://github.com/awesomeWM/awesome/commit/4c322a613f508180faf0dc3049f35b2744e85905) could be one reason).

Not only does it only hang with the external monitor, it also only hangs with AwesomeWM 4.3. For me, it's currently more plausible that this is something to do with `io.popen` that may have been changed and now malfunctions when the external monitor is connected, rather than that it's something to do with the audio devices specifically - if I disable the volume widget but enable the caps lock widget (which also uses `io.popen`), it also hangs. So the volume widget may only cause problems because it happens to call `io.popen`.

As @Elv13  points out, if `io.popen` isn't/can't be supported, then maybe that's "ok".

**Comment by psychon** [comment]
Just in case I search for it again: The caps lock widget just runs `xset q | grep Caps | cut -d ' ' -f 10`.

Another random idea: Does adding `awesome.sync()` somewhere close to the call to `io.popen` help? You write that `popen` hangs if run between `hints.init()` and the initialisation of the volume widget. What if you add `awesome.sync()` just before the call to the volume widget?

(The theory here is: Perhaps the X11 server is currently grabbed by AwesomeWM and so other processes trying to talk to the X11 server block? Awesome grabs the server in some places, but is careful to do so only for short times and not to run any Lua code in between. However, it could be that XCB's output buffer is filled up after `xcb_grab_server` and `xcb_ungrab_server`. That would mean that the grab is sent to the X11 server, but the ungrab is still waiting in XCB's output buffer. `awesome.sync()` actually does something else, but as a side effect it causes the output buffer to be flushed.

This would also explain the weird "it only happens between here and here":  The bug would depend on how many X11 requests are sent and creating new wiboxes definitely creates a number of requests.)

**Comment by furgerf** [comment]
Thanks for the comments, interesting! Before I forget again: In case of a freeze,  I need to `pkill -9 awesome` to get it to restart in case that's helpful.

> This would also explain the weird "it only happens between here and here": The bug would depend on how many X11 requests are sent and creating new wiboxes definitely creates a number of requests.)

You're right here - I hadn't considered that the commands would work again at a later point. But that's the case, after the widgets are created, things like `amixer` or `xset` don't cause freezes anymore.

> Another random idea: Does adding `awesome.sync()` somewhere close to the call to `io.popen` help? You write that `popen` hangs if run between `hints.init()` and the initialisation of the volume widget. What if you add `awesome.sync()` just before the call to the volume widget?

It does! Both for custom `io.popen` and for creating the widgets.

What does that mean now? I'm afraid I don't think I've heard of XCB before... From what I do understand, is it possible that I'm seeing these freezes only in 4.3 because the X server is now grabbed more often/longer?

**Comment by psychon** [comment]
> I'm afraid I don't think I've heard of XCB before...

XCB stands for "X11 C bindings". It is the library awesomeWM uses for talking to the X server. And for performance reason, XCB does not send requests to the X11 server immediately, but only when explicitly told to or when its buffer is full. Thus, random changes can make this bug appear or disappear. Just sending less X11 requests somewhere might mean that the buffer is not flushed "in the middle of something". Or sending more X11 requests could move the time when the buffer is full sufficiently far ahead so that again the grab and ungrab are not sent separately.

> From what I do understand, is it possible that I'm seeing these freezes only in 4.3 because the X server is now grabbed more often/longer?

Not really. In fact, from the top of my head, 4.3 and 4.2 use grabbing in the same number of places. It's just that "the buffer is full" is basically happening at random times.

I'm not quite sure what to do about this bug in awesome. We could just flush whenever we generate an UngrabServer request to make sure it is sent out as soon as possible...

**Comment by Elv13** [comment]
>  We could just flush whenever we generate an UngrabServer request to make sure it is sent out as soon as possible...

While it would solve *this very specific* instance of the bug, it would not fix the infinite number of other ways to trigger it (most of them unfixable). IMHO this falls into the `--no-argb` category of "this is someone else problem, mitigating it vigorously will make everything worst/slower/less_maintainable". We should probably start to deprecate (monkeypatch) `io.popen` (as in, deprecate now, not in v5.0) then show a warning dialog (once) in v5.x.

**Comment by psychon** [comment]
> We should probably start to deprecate (monkeypatch) `io.pope`

I briefly looked into doing that, but it would cause `awful.completion` to spew deprecation warnings.

**Comment by Elv13** [comment]
Closing this since it is "resolved". Deprecating `io.popen` can be done in a new issue.
